src/_pytest/pytester.py

```diff
@@ -1061,7 +1065,27 @@ def run(self, *cmdargs):
             popen = self.popen(
                 cmdargs, stdout=f1, stderr=f2, close_fds=(sys.platform != "win32")
             )
-            ret = popen.wait()
+            timeout = kwargs.get("timeout")
+            if timeout is None:
+                ret = popen.wait()
+            elif six.PY3:
+                try:
+                    ret = popen.wait(timeout)
+                except subprocess.TimeoutExpired:
+                    raise self.TimeoutExpired()
+            else:
+                end = monotonic.monotonic() + timeout
+
+                while True:
+                    ret = popen.poll()
+                    if ret is not None:
+                        break
+
+                    remaining = end - monotonic.monotonic()
+                    if remaining <= 0:
+                        raise self.TimeoutExpired()
```

Comment:
I suggest that we add a message here with the command and how many seconds was the timeout.